s4e-backend/src/main/java/pl/cyfronet/s4e/controller/PasswordController.java

```diff
@@ -0,0 +1,116 @@
+            @ApiResponse(code = 200, message = "Password reset token is valid"),
+            @ApiResponse(code = 401, message = "The token has expired"),
+            @ApiResponse(code = 404, message = "The token was not found")
+    })
+    @GetMapping(value = "/token-validate")
+    public ResponseEntity<?> validateToken(@RequestParam @NotEmpty @Valid String token) throws NotFoundException, PasswordResetTokenExpiredException {
+        val resetToken = passwordService.findByToken(token)
+                .orElseThrow(() -> new NotFoundException("Provided token '" + token + "' not found"));
+
+        if (resetToken.getExpiryTimestamp().isBefore(LocalDateTime.now())) {
+            throw new PasswordResetTokenExpiredException("Provided token '" + token + "' expired");
+        }
+
+        return ResponseEntity.ok().build();
+    }
+
+    @ApiOperation("Reset password")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "Password reset was successful"),
+            @ApiResponse(code = 400, message = "Passwords were incorrect"),
+            @ApiResponse(code = 404, message = "The token was not found")
+    })
+    @PostMapping(value = "/password-reset")
+    @ResponseBody
+    public ResponseEntity<?> resetPassword(@RequestBody @Valid PasswordResetRequest passwordReset, @RequestParam("token") String token) throws NotFoundException, BadRequestException {
+        val optionalResetPasswordToken = passwordService.findByToken(token)
+                .orElseThrow(() -> new NotFoundException("Provided token '" + token + "' not found"));
+        if (optionalResetPasswordToken.getAppUser().getPassword().equals(passwordReset.getOldPassword())) {
+            AppUser user = optionalResetPasswordToken.getAppUser();
+            user.setPassword(passwordReset.getNewPassword());
+            appUserService.update(user);
+            passwordService.delete(optionalResetPasswordToken.getId());
+        } else {
+            throw new BadRequestException("Provided passwords were incorrect");
+        }
+
+        return ResponseEntity.ok().build();
+    }
+
+    @ApiOperation("Change password")
+    @ApiResponses({
+            @ApiResponse(code = 200, message = "Password change was successful"),
+            @ApiResponse(code = 400, message = "Passwords were incorrect"),
+            @ApiResponse(code = 404, message = "The user was not found")
+    })
+    @PostMapping(value = "/password-change")
+    @ResponseBody
+    @PreAuthorize("!isAnonymous()")
+    public ResponseEntity<?> changePassword(@RequestBody @Valid PasswordResetRequest passwordReset) throws NotFoundException, BadRequestException {
+        val optionalAppUser = appUserService.findByEmail(
+                ((AppUserDetails) SecurityContextHolder.getContext().getAuthentication().getDetails()).getAppUser().getEmail())
+                .orElseThrow(() -> new NotFoundException());
+        if (optionalAppUser.getPassword().equals(passwordReset.getOldPassword())) {
```

Comment:
Actually, this check will always fail. The password in AppUser is hashed. Use PasswordEncoder to encode the password from PasswordResetRequest before checking for equality.